cluster/src/main/java/io/servicefabric/cluster/Cluster.java

```diff
@@ -121,26 +118,32 @@ public static ICluster newInstance(ClusterConfiguration config) {
     clusterMembership.setSyncGroup(config.clusterMembershipSettings.getSyncGroup());
 
     // Build cluster component
-    ICluster cluster = new Cluster(transport, failureDetector, gossipProtocol, clusterMembership);
-
-    // Auto start
-    if (config.autoJoin) {
-      cluster.join();
-    }
+    Cluster cluster = new Cluster(transport, failureDetector, gossipProtocol, clusterMembership);
 
     return cluster;
   }
 
   @Override
-  public ITransportChannel to(ClusterMember member) {
+  public void send(ClusterMember member, Message message) {
+    checkJoinedState();
+    transport.send(member.endpoint(), message);
+  }
+
+  @Override
+  public void send(ClusterMember member, Message message, SettableFuture<Void> promise) {
     checkJoinedState();
-    return transport.to(member.endpoint().endpoint());
+    transport.send(member.endpoint(), message, promise);
   }
 
   @Override
-  public Observable<TransportMessage> listen() {
+  public Observable<ClusterMessage> listen() {
     checkJoinedState();
-    return transport.listen();
+    return transport.listen().map(new Func1<TransportMessage, ClusterMessage>() {
+      @Override
+      public ClusterMessage call(TransportMessage transportMessage) {
+        return new ClusterMessage(transportMessage.message(), membership().member(transportMessage.endpoint().id()));
```

Comment:
This is dangerous, because membership.member theoretically can return null and constructor of ClusterMessage will throw IllegalArgumentException